[DataGrid] Add quick filtering engine (#4317)
mui/mui-x: docs/data/data-grid/filtering/filtering.md

## Quick filter

The package also provides a component `<GridToolbarQuickFilter/>` which filters rows without specifying the concerned columns.
By default, the quick filter considers the input as a list of values separated by space and keeps only rows that contain all the values.

{{"demo": "QuickFilteringGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Modifying the logic

The logic quick filter logic can be switched to filter rows that contain _at least_ one value instead for each of them.
To do so, set `quickFilterLogic` equals to `GridQuickFilterLogic.Or` as follow:

```js
initialState={{
  filterModel: {
    quickFilterLogic: GridQuickFilterLogic.Or
  }
}}
```

> The quick filter will only consider columns with types `string` or `number`

Reviewer: Could it support `singleSelect` ?

And below you talk about a `date` column.
Are `string` and `number` the only supported by default ? If so, maybe this sentence should be reworded.

Author: Yes, single-select could be supported. I wanted first to validate the global solution before doing this particular case. About the date, there is quite a lot of way to specify a date in a text area

## Quick filter

> The package also provides a component `<GridToolbarQuickFilter/>` which filters rows without specifying the concerned columns.

Reviewer 2: I think we could first introduce what quick filter is.

```suggestion
Quick filter allows to filter rows using free text.
To enable it, you can add the `<GridToolbarQuickFilter />` component to the toolbar.
```

By default, the quick filter considers the input as a list of values separated by space and keeps only rows that contain all the values.

{{"demo": "QuickFilteringGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Custom filtering logic

> The logic quick filter logic can be switched to filter rows that contain _at least_ one value instead for each of them.

Reviewer 2: ```suggestion
The logic used for quick filter can be switched to filter rows that contain _at least_ one of the values specified, instead of testing if it contains all of them.
```

> To do so, set `quickFilterLinkOperator` equals to `GridLinkOperator.Or` as follow:

Reviewer 2: ```suggestion
To do so, set `quickFilterLinkOperator` to `GridLinkOperator.Or` as follow:
```

```js
initialState={{
  filterModel: {
    quickFilterLinkOperator: GridLinkOperator.Or
  }
}}
```

> With default settings, the quick filter will only consider columns with types `string`,`number` and `singleSelect`

Reviewer 2: ```suggestion
With the default settings, quick filter will only consider columns with types `string`, `number` and `singleSelect`.
```

> - For columns with type `string` the cell must contain the value to pass the filter
> - For columns with type `number` the cell must equal the value to pass the filter
> - For columns with type `singleSelect` the cell must start like the value to pass the filter

Reviewer 2: I didn't understand what "start like the value" meant.

```suggestion
- For columns with type `string`, the cell must contain the value to pass the filter
- For columns with type `number`, the cell must be equal to the value to pass the filter
- For columns with type `singleSelect`, the cell must start like the value to pass the filter
```

To modify or add the quick filter operators, add the property `getApplyQuickFilterFn` to the column definition.
This function is quite similar to `getApplyFilterFn`.
This function takes as an input a value of the quick filter and returns another function that takes the cell value as an input and returns `true` if it satisfies the operator condition.

> In the example bellow, we filter a `date` column by checking if it contains the correct year.

Reviewer 2: No "we".

```suggestion
In the example below, a custom filter is created for the `date` column to check if it contains the correct year.
```

## Quick filter

> Quick filter allows filtering rows using free text.

Reviewer 3: ```suggestion
Quick filter allows filtering rows by multiple columns with a single text input.
```

To enable it, you can add the `<GridToolbarQuickFilter />` component to your custom toolbar or pass `showQuickFilter` to the default `<GridToolbar />`.

By default, the quick filter considers the input as a list of values separated by space and keeps only rows that contain all the values.

{{"demo": "QuickFilteringGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Custom filtering logic

The logic used for quick filter can be switched to filter rows that contain _at least_ one of the value specified instead of testing if it contains all of them.
To do so, set `quickFilterLogicOperator` to `GridLinkOperator.Or` as follow:

```js
initialState={{
  filterModel: {
    quickFilterLogicOperator: GridLinkOperator.Or
  }
}}
```

With the default settings, quick filter will only consider columns with types `string`,`number`, and `singleSelect`.

- For columns with type `string`, the cell must contain the value to pass the filter

> - For columns with type `number`, the cell must be equal to the value to pass the filter

Reviewer 3: ```suggestion
- For `number` columns, the cell must be **equal** any of the values
```

> - For columns with type `singleSelect`, the cell must start like the value to pass the filter

Reviewer 3: ```suggestion
- For `singleSelect`columns , the cell must **start with** any of the values
```

## Quick filter

Quick filter allows filtering rows by multiple columns with a single text input.
To enable it, you can add the `<GridToolbarQuickFilter />` component to your custom toolbar or pass `showQuickFilter` to the default `<GridToolbar />`.

By default, the quick filter considers the input as a list of values separated by space and keeps only rows that contain all the values.

{{"demo": "QuickFilteringGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Custom filtering logic

The logic used for quick filter can be switched to filter rows that contain _at least_ one of the values specified instead of testing if it contains all of them.
To do so, set `quickFilterLogicOperator` to `GridLinkOperator.Or` as follow:

```js

> initialState={{
>   filterModel: {
>     quickFilterLogicOperator: GridLinkOperator.Or
>   }
> }}

Reviewer 2: Without `items` it crashes.

```suggestion
initialState={{
  filter: {
    filterModel: {
      items: [],
      quickFilterLogicOperator: GridLinkOperator.Or,
    },
  },
}}
```
